### ISSUE TITLE
Container cost and size configurability

### ISSUE THREAD

**Issue by nbusseneau** [MSchmoecker/Dynamic-Storage-Piles#2]
Hey, thanks for your mods and Jotunn 😄 

What do you think about adding configurability for the container sizes and initial cost? I don't think allowing to configure each stack individually is needed, but just setting the shared cost/size of all containers.

Use case: skimming through the description I actually expected the stacks would not allow for storing more than vanilla (e.g. 1 full wood stack = 50 wood). If the container size was configurable, I would personally set it to just 1 slot so as to get to that, instead of allowing to store so many resources in one stack 😅 

And with only one slot, allowing to create a stack with a cost of just 1 instead of 10 would make it less restricting.

**Comment by MSchmoecker** [comment]
Thanks!

First of all, let me explain the reasoning behind the choice. The stacks have the same inventory space and cost the same amount of items as a wooden chest. However, they usually take up more build space and only accept one item type. So I think it's a fair trade.

That being said, you can use mods like [AzuContainerSizes](https://thunderstore.io/c/valheim/p/Azumatt/AzuContainerSizes/) to change the inventory sizes and [WackysDatabase](https://thunderstore.io/c/valheim/p/WackyMole/WackysDatabase/) to change the piece cost.
There will be no configs for the time being, as the added complexity doesn't seem to be justified when the other mods already do the job very well.

The prefab names can be found here (or use a mod like VNEI or AzuHoverStats) https://github.com/MSchmoecker/Dynamic-Storage-Piles/blob/e1fe35f93d7c2dcb0da7e24d0f1cc016f9707b23/DynamicStoragePiles/DynamicStoragePiles.cs#L52-L63

**Comment by nbusseneau** [comment]
Thanks, yeah I know about those but believe it's better when the mod does it directly (especially because of how WackyDB cloning behaves) 😅 

Is this a soft or hard stance (i.e. you would not do it yourself but would you accept or refuse a pull request adding this)?

**Comment by MSchmoecker** [comment]
If a PR adds this feature, works good and isn't too complex, it will very likely be merged.

This implies the crafting cost is synced with Jotunn's ConfigSync and all edge cases regarding inventory size changes are handled properly. E.g when the size is set smaller and items are still inside the pile

**Comment by nbusseneau** [comment]
Aye, maybe I will try something at some point then.

**Comment by nbusseneau** [comment]
FYI, in the end I do not intend to try and submit a PR:

- After thinking about it some more, your rationale makes sense. Considering the vanilla chest cost and container size, it indeed is necessary to have something in line with this for Dynamic Storage Piles to have a sensible use case in game.
- After testing as part of our modpack, we ultimately decided to go with https://thunderstore.io/c/valheim/p/Balrond/balrond_containers/ (mostly because it fit the camp vibe we wanted to go for better, e.g. with the generic wood shed allowing to stockpile multiple types of wood).

So instead I am currently making changes to Balrond Containers to have it work in a way similar to Dynamic Storage Piles (notably adding in restrictions on allowed items, because by default there is none).

I suppose we can close this issue, unless you want to leave it up open for someone else to maybe grab in the future 😄

**Comment by Ujhik** [comment]
I'm working on container size configurability.

@MSchmoecker do you think it's good enough if before loading the container, if it has any item inside, it sets the size to 2x5 if the customized size is smaller, otherwise it sets the size to the customized one? or do you think is better making the items to group together in the minimum number of stacks, then checking if the number of stacks is bigger than the custom size and if it's not putting them to the topLeft position?

I would gladly implement any better idea, already have the working code for the first question with a HarmonyPatch prefix to container Load to resize it if necessary

**Comment by ZenDragonX** [comment]
I just wanted to chime in here that I think container size configurability is a great idea!  

I also expected it to be 1x1 with the cost of a vanilla stack.  It could default to be being filled with the cost of the initial stack.

Example:  
- A vanilla stack of wood costs 50 wood and has a value of 50 wood when destroyed. When built it looks like a full stack.  
- A dynamic stack costs 10 wood and is empty.  The cursor looks like a full stack but when placed looks like just a single log.  Visually it's weird.  I would change that to:

How I would expect it to behave:
- The dynamic stack would cost 50 wood (same as vanilla) and come pre-filled with 50 wood already in it with a single 1x1 slot.
- When placed would be a full stack. Therefor it would look seamless when compared to vanilla both when building and expected behavior.

Bonus:
An ease of use nice to have config option would be to hide the vanilla stacks from being on the build menu.  Yes, WackyDB is doable, but pain in the ass compared to a checkbox in the configs.

**Comment by MSchmoecker** [comment]
@Ujhik, yeah checking the items before loading the container is a good solution, although it should set the inventory to the smallest possible space and not default. I.e. max(wanted, config). The inventory size can be changed after the container has loaded as far as I know, so subscribing to the inventory change event and checking if the inventory size should be changed is possible.

@ZenDragonX, the pre-filling is an interesting idea, didn't thought about it so far. There is already a config to disable placing of vanilla stacks
